pep-9999.rst

```diff
@@ -0,0 +1,180 @@
+PEP: 9999
+Title: Callable Modules
+Author: Amethyst Reese <amethyst at n7.gg>
+Sponsor: Łukasz Langa <lukasz at python.org>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 20-Apr-2023
+Python-Version: 3.12
+
+
+Abstract
+========
+
+Modules are currently not directly callable. Classes can define a ``__call__``
+method that makes instance objects callable, but defining a similarly named
+function in the global module scope has no effect, and that function can
+only be called by importing or referencing it directly as ``module.__call__``.
+:pep:`562` added support for ``__getattr__`` and ``__dir__`` for modules, but
+defining ``__getattr__`` to return a value for ``__call__`` still does not
+make a module callable.
+
+This PEP proposes support for a ``__call__`` function defined on modules
+to make the module object itself callable.
+
+
+Motivation
+==========
+
+Many modules have only a single primary interface to their functionality.
+In many cases, that interface is a single callable object, where being able
+to use a simple import and using the module as a callable provides a more
+"pythonic" interface for users:
+
+.. code-block:: python
+
+    # user.py
+
+    import fancy
+
+    @fancy
+    def func(...):
+        ...
+
+Currently, providing this style of interface requires modifying the module
+object at runtime to make it callable.
+
+This is commonly done by replacing the module object in ``sys.modules`` with
+a callable alternative (such as a function or class instance):
+
+.. code-block:: python
+
+    # fancy.py
+
+    def fancy(...):
+        ...
+
+    sys.modules[__name__] = fancy
+
+This has the effect of making the original module effectively unreachable
+without further hooks from the author, even with ``from module import member``.
+It also results in a "module" object that is missing all of the special module
+attributes, including ``__doc__``, ``__package__``, ``__path__``, etc.
+
+Alteratively, a module author can choose to override the module's ``__class__``
+property with a custom type that provides a callable interface:
+
+.. code-block:: python
+
+    # fancy.py
+
+    def fancy(...):
+        ...
+
+    class FancyModule(types.ModuleType):
+        def __call__(self, ...):
+            return fancy(...)
+
+    sys.modules[__name__].__class__ = FancyModule
+
+The downside of either approach is that it not only results in extra
+boilerplate, but also results in type checker failures because they don't
+recognize that the module is callable at runtime:
+
+.. code-block:: console
+
+    $ mypy user.py
+    user.py:3: error: Module not callable  [operator]
+    Found 1 error in 1 file (checked 1 source file)
+
+
+Specification
+=============
+
+When a module object is called, and a ``__call__`` function is found (either
```

Comment:
Here (and in the final para of the Abstract & final para of Rejected Ideas), you reference the ‘`__call__` function’, but just below this line in the specification you show how an arbitrary callable (e.g. a class) can be assigned to the `__call__` name. Two resultant questions:

1. At potential risk of redundancy, does it make sense to use ‘`__call__` callable’ (or object) as the phrase here?
2. Should greater attention be drawn to the ‘arbitrary callable’ case? I admit I was a little surprised on realising you supported this!

A

Reply:
Clarified the intent to allow arbitrary callable objects to the module's `__call__` property.